protocols/tls_cryptolib/src/lib.rs

```diff
@@ -0,0 +1,529 @@
+                let seq2 = seq2.slice(1, seq2.len() - 1);
+                if element_type == 0x30u8 {
+                    let len_len = get_length_length(&seq2);
+                    if len_len == 0 {
+                        let oid_len = get_short_length(&seq2);
+                        if oid_len >= 9 {
+                            // ecPublicKey oid incl tag: 06 07 2A 86 48 CE 3D 02 01
+                            // FIXME: This shouldn't be necessary. Instead public_byte_seq!
+                            //        should be added to the typechecker.
+                            let expected = ByteSeq::from_seq(&EcOidTag(secret_bytes!([
+                                0x06u8, 0x07u8, 0x2Au8, 0x86u8, 0x48u8, 0xCEu8, 0x3Du8, 0x02u8, 0x01u8
+                            ])));
+                            let oid = seq2.slice(1, 9);
+                            let mut ec_pk_oid = true;
+                            for i in 0..9 {
+                                let oid_byte_equal = U8::declassify(oid[i]) == U8::declassify(expected[i]);
+                                ec_pk_oid = ec_pk_oid && oid_byte_equal;
+                            }
+                            if ec_pk_oid {
+                                // We have an ecPublicKey, skip the inner sequences
+                                // and read the public key from the bit string
+                                let bit_string = seq2.slice(oid_len + 1, seq2.len() - oid_len - 1);
+                                // We only support uncompressed points
+                                if U8::declassify(bit_string[0]) == 0x03u8 {
+                                    let pk_len = declassify_usize_from_U8(bit_string[1]); // 42
+                                    let _zeroes = declassify_usize_from_U8(bit_string[2]); // 00
+                                    let _uncompressed = declassify_usize_from_U8(bit_string[3]); // 04
+                                    pk = bit_string.slice(4, pk_len - 2);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            seq1 = seq1.slice(len, seq1.len() - len);
+        }
+    }
+    if pk.len() == 0 {
+        Result::<VerificationKey, CryptoError>::Err(CryptoError::InvalidCert)
+    } else {
+        Result::<VerificationKey, CryptoError>::Ok(pk)
+    }
+}
+
+fn concat_signature(r: P256Scalar, s: P256Scalar) -> Result<Signature, CryptoError> {
+    let signature = Signature::new(0)
+        .concat(&r.to_byte_seq_be())
+        .concat(&s.to_byte_seq_be());
+        Result::<Signature, CryptoError>::Ok(signature)
+}
+
+fn p256_sign(ps: &SignatureKey, payload: &ByteSeq, ent: Entropy) -> Result<Signature, CryptoError> {
+    let random = Random::from_seq(&ent.slice_range(0..32));
+    // FIXME: return an error when nonce is not a valid scalar value.
+    let nonce = P256Scalar::from_byte_seq_be(&random);
```

Comment:
Indeed. If you look at the type declaration:

```rust

public_nat_mod!(
    type_name: P256Scalar,
    type_of_canvas: ScalarCanvas,
    bit_size_of_field: 256, // XXX: Unfortunately we can't use constants here.
    modulo_value: "ffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551"
);
```
What you really want is 

```rust
let nonce = ScalarCanvas::from_byte_seq_be(&random);
let nonce = P256Scalar::from_canvas(nonce)?;
```
What is missing now is a `from_canvas` function that may error right?

Reply:
The problem is that `from_canvas` just forces the value into the scalar. Instead we have to check that the value is in `[1, order-1]` and return an error if not. We could do this in the P256 code. But it's a pretty common functionality such that I think it would make sense to have it in the library.

Reply:
Yeah I would agree that we should change the return type of `from_canvas` to a `Result` :)